### PR TITLE
[Templates] SPA template proxy approach update

### PR DESCRIPTION
**Pull request by javiercn**
* Introduces a new package Microsoft.AspNetCore.SpaProxy to handle launching the SPA proxy as the frontend for the app.
* Removes unnecessary references to Microsoft.AspNetCore.SpaExtensions.
* Updates the angular template to use a new approach where the SPA frontend proxy forwards requests to the backend API.

Additional details are [here](https://gist.github.com/javiercn/b28cb0ea0912de8313e0fd0cff9021d9#file-spec-md)

**TL;DR**
There is a new package responsible for launching an NPM command to start the proxy upon startup.
The proxy launches using https in 5002 and forwards requests to 5001 for the configured paths.
Users can run from VS with a single F5 gesture (same experience).
Users can run with dotnet run with a single gesture.
Users can start the proxy manually with npm start and then do dotnet run separately.

### CODE REVIEW

**Review by pranavkm — APPROVED** [review]

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                // it makes sure that we don't send lots of requests and that we give the process some time to
+                // get up and running
+                await Task.Delay(1000);
+                var response = await _httpClient.GetAsync(_options.ServerUrl, cancellationToken);
+                var running = response.IsSuccessStatusCode;
+                return running;
+            }
+            catch (HttpRequestException httpException)
+            {
+                _logger.LogDebug(httpException, "Failed to reach the SPA Development proxy.");
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:66]: This is only used for tooling purposes at development time, the LoggerMessage pattern is a bit overkill for this case.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                        // we now start a proxy every time.
+                        // We can't guarantee that we stop/cleanup the proxy on every situation (for example if someone)
+                        // kills this process in a "rude" way, but this gets 95% there.
+                        // For cases where the proxy is left open and where there might not be a "visible" window the recomendation
+                        // is to kill the process manually. (We will not fail, we will simply notify the proxy is "already" up.
+                        if (!_spaProcess.CloseMainWindow())
+                        {
+                            _spaProcess.Kill(entireProcessTree: true);
+                            _spaProcess = null;
+                        }
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:173]: This have made me think that this code might not be 100% correct, since I'm not 100% sure I should mock with the `_spaProcess` instance when running in the context of the finalizer. That said, I don't think that matters since this object is managed by the DI container and will always be disposed, and the worst case scenario is an that an exception is thrown and caught within the finalizer, which I think it's fine.

I'm not going to change it since that would require me to retest all the templates across OSs again.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+namespace Microsoft.AspNetCore.SpaProxy
+{
+    internal class SpaProxyLaunchManager : IHostedService, IDisposable
+    {
+        private readonly SpaDevelopmentServerOptions _options;
+        private readonly ILogger<SpaProxyLaunchManager> _logger;
+        private readonly HttpClient _httpClient = new(new HttpClientHandler()
+        {
+            // It's ok for us to do this here since this service is only plugged in during development.
+            ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
```
Comment (SteveSandersonMS) [inline src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:25]: This is a fairly minor point as it would only matter potentially in the future but... it feels kinda dangerous to have this `_httpClient` lying around in this state, as someone might start using it for other purposes in the future where they actually need cert validation to happen as usual.

Since this `HttpClient` instance is only used to probe for the SPA server startup, it might be nice not to have it as a class-level field, and instead new up an instance each time we're going to do a set of calls into `ProbeSpaDevelopmentServerUrl`. The perf cost of doing that would be nothing in practice.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                _logger.LogInformation($"No SPA development server running at {_options.ServerUrl} found.");
+                await StartSpaProcessAndProbeForLiveness(cancellationToken);
+            }
+        }
+
+        private async Task<bool> ProbeSpaDevelopmentServerUrl(CancellationToken cancellationToken)
+        {
+            try
+            {
+                var response = await _httpClient.GetAsync(_options.ServerUrl, cancellationToken);
```
Comment (SteveSandersonMS) [inline src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:60]: How long might this call last? Do we generally find that the SPA dev servers reply instantly if they are alive, or do they potentially wait minutes while going through a webpack build?

Maybe even if they do take ages it's desirable to wait for them. Just asking to clear up my understanding of the startup flow.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            }
+        }
+
+        private async Task StartSpaProcessAndProbeForLiveness(CancellationToken cancellationToken)
+        {
+            LaunchDevelopmentProxy();
+            var sw = Stopwatch.StartNew();
+            var livenessProbeSucceeded = false;
+            var maxTimeoutReached = false;
+            while (_spaProcess != null && !_spaProcess.HasExited && !livenessProbeSucceeded && !maxTimeoutReached)
```
Comment (SteveSandersonMS) [inline src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:77]: Looks like the `!livenessProbeSucceeded` part of this condition isn't needed, because it only becomes true on L79, after which we'd break out of the loop anyway.

```suggestion
            while (_spaProcess != null && !_spaProcess.HasExited && !maxTimeoutReached)
```

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                livenessProbeSucceeded = await ProbeSpaDevelopmentServerUrl(cancellationToken);
```
Comment (SteveSandersonMS) [inline src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:79]: Should we impose a shorter timeout here than whatever `cancellationToken` has? Since we're polling, it would seem more natural to only wait a few seconds at a time, rather than potentially waiting for minutes on each loop.

The `MaxTimeout` will only be honored with reasonable accuracy if each polling loop is short enough.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            {
+                _logger.LogInformation($"SPA development server running at '{_options.ServerUrl}'");
+            }
+        }
+
+        private void LaunchDevelopmentProxy()
+        {
+            try
+            {
+                var space = _options.LaunchCommand.IndexOf(' ');
```
[the remark this answers is not in the document]
Comment (SteveSandersonMS) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:107]: We might also want to note that we're deliberately not supporting `LaunchCommand` values that contain spaces (for example: `"c:\Users\My User\nodejs\bin\npm.cmd"`, which might seem valid given the surrounding quotes).

Somebody out there is going to run into trouble with spaces but it's probably enough of an edge case not to address if it requires bringing in some more complex parsing logic.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                        // and the service keeps running.
+                        // Compared to performing Ctrl+C on the window or closing the window for the newly spawned
+                        // process which seems to do the right thing.
+                        // Process.CloseMainWindow seems to do the right thing in this situation and is doable since
+                        // we now start a proxy every time.
+                        // We can't guarantee that we stop/cleanup the proxy on every situation (for example if someone)
+                        // kills this process in a "rude" way, but this gets 95% there.
+                        // For cases where the proxy is left open and where there might not be a "visible" window the recomendation
+                        // is to kill the process manually. (We will not fail, we will simply notify the proxy is "already" up.
+                        if (!_spaProcess.CloseMainWindow())
```
[the remark this answers is not in the document]
Comment (SteveSandersonMS) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:169]: I honestly have no idea what combination of behaviors might occur for different SPA dev servers and OSes in which shutdown scenarios. From memory, it's very tricky trying to get consistent behaviors for shutting down child processes.

In all likelihood we'll just have to see what happens in practice with the React/Angular/etc dev servers on different OSes and whether it leads to a workflow people find convenient. We can tweak the shutdown logic later if there are any problematic cases, such as something being left half-shut-down.

File: src/Middleware/Spa/SpaProxy/src/build/Microsoft.AspNetCore.SpaProxy.targets [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+      <_SpaProxyServerLaunchConfigLines Include="    &quot;MaxTimeoutInSeconds&quot;: &quot;$(SpaProxyTimeoutInSeconds)&quot;" />
+      <_SpaProxyServerLaunchConfigLines Include="  }" />
+      <_SpaProxyServerLaunchConfigLines Include="}" />
+    </ItemGroup>
+    <WriteLinesToFile File="$(_SpaProxyServerLaunchConfig)" Lines="@(_SpaProxyServerLaunchConfigLines)" WriteOnlyWhenDifferent="true" Overwrite="true" />
+    <ItemGroup>
+      <ContentWithTargetPath Include="$(_SpaProxyServerLaunchConfig)" CopyToOutputDirectory="PreserveNewest" CopyToPublishDirectory="Never" TargetPath="spa.proxy.json" />
+      <FileWrites Include="$(_SpaProxyServerLaunchConfig)" />
+    </ItemGroup>
+  </Target>
```
Comment (SteveSandersonMS) [inline src/Middleware/Spa/SpaProxy/src/build/Microsoft.AspNetCore.SpaProxy.targets:24]: This is an interesting design. TBH I was expecting developers to create and edit the `spa.proxy.json` file themselves - I didn't predict we'd be generating it at build time based on a set of MSBuild properties.

One drawback to generating it like this is with formatting. The logic here will produce an invalid JSON file if `SpaProxyLaunchCommand` contains a `"` character, which could actually be very common since people like to pass all kinds of arguments to their startup commands. Not sure if `\n` or other characters would cause issues too.

Do you think the MSBuild properties are more discoverable/convenient than just telling people to drop in a `spa.proxy.json` file to their project? In both cases they will only know about this by reading some docs or tutorial, so  in either design we're relying on them copy/pasting stuff.

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      '--export-path',
+      certFilePath,
+      '--format',
+      'Pem',
+      '--no-password',
+    ],
+    {
+      stdio: 'inherit',
+    }
+  ).on('exit', (code) => process.exit(code));
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js:39]: By all means please ignore this suggestion if you don't like it, but the code might seem less intimidating and easier to skim-read if formatted closer to the JS-typical brace style and more compactly:

```suggestion
  spawn('dotnet', [
    'dev-certs', 'https',
    '--no-password',
    '--export-path', certFilePath,
    '--format', 'Pem',
  ], { stdio: 'inherit' })
  .on('exit', code => process.exit(code));
```

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+const baseFolder =
+  process.env.APPDATA !== undefined && process.env.APPDATA !== ''
+    ? `${process.env.APPDATA}/ASP.NET/https`
+    : `${process.env.HOME}/.aspnet/https`;
+
+const certificateName = process.argv.map(arg => arg.match(/--name=(?<value>.+)/i))
+                                    .filter(Boolean)
+                                    .reduce((previous, current) => previous || current.groups.value, undefined) ||
+  process.env.npm_package_name;
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js:14]: Some slightly funky formatting going on here. Also if you're interested, this can be done without `reduce` which would reduce the scariness factor a bit:

```suggestion
const certificateArg = process.argv.map(arg => arg.match(/--name=(?<value>.+)/i)).filter(Boolean)[0];
const certificateName = certificateArg ? certificateArg.groups.value : process.env.npm_package_name;
```

The ternary could be reduced further to `certificateArg?.groups.value || process.env.npm_package_name` but I think that's harder to understand.

Since this is just code style please feel free to ignore the suggestion!

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/Startup.cs [hunk 12]
```diff
@@ -112,19 +101,8 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
-                // To learn more about options for serving an Angular SPA from ASP.NET Core,
-                // see https://go.microsoft.com/fwlink/?linkid=864501
-
-                spa.Options.SourcePath = "ClientApp";
 
-                if (env.IsDevelopment())
-                {
-                    spa.UseAngularCliServer(npmScript: "start");
-                }
+                endpoints.MapFallbackToFile("index.html");
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/Startup.cs:105]: Well, this file looks a lot nicer now! Great stuff!

**Review by SteveSandersonMS — APPROVED** [review]
This looks superb to me. I'm impressed you were able to find ways of bringing so much functionality while hardly increasing the conceptual load or implementation complexity at all.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:107]: This only needs to support `yarn/npm <<verb>> <<args>>` if you want something very specialize then you can always wrap it on a shell script and use that as your command.

[the remark this answers is not in the document]
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:107]: I added a comment to make it clear.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:169]: All the frameworks use the webpack dev server in the end. From my empirical tests `CloseMainWindow()` tries to do the right thing or returns false (but conveniently doesn't throw).

The behavior here is at least comparable to the behavior in the script runner SpaExtensions, I verified the shutdown across OSes on common scenarios (ctrl+c, close window, etc.). We might end up not even shipping this package in RTM, so I think it's good for now (or at least as good as its going to get)

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs (hunk 4 above)
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:60]: It comes back right away. However I've updated the code to add a timeout after 1 second or if the host request cancellation, whatever happens first.

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs (hunk 6 above)
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:79]: Each poll attempt will wait at most 1 second to receive an answer and 1 second after each unsuccessful connection

File: src/Middleware/Spa/SpaProxy/src/build/Microsoft.AspNetCore.SpaProxy.targets (hunk 9 above)
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/build/Microsoft.AspNetCore.SpaProxy.targets:24]: > One drawback to generating it like this is with formatting. The logic here will produce an invalid JSON file if `SpaProxyLaunchCommand` contains a `"` character, which could actually be very common since people like to pass all kinds of arguments to their startup commands. Not sure if `\n` or other characters would cause issues too.

In general you should put arguments inside `package.json` otherwise you loose the ability to run `npm/yarn start` from the cli. I started down that route where the command in the project file contained everything. There are alternatives to make this work for all people if needed, like wrapping the command in a script file and using the script for launch instead.

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js (hunk 10 above)
Comment (javiercn) [reply src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js:39]: vs code formatting is fighting me :). I'll fix it, it used to be that way

File: src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js (hunk 11 above)
Comment (javiercn) [reply src/ProjectTemplates/Web.Spa.ProjectTemplates/content/Angular-CSharp/ClientApp/aspnetcore-https.js:14]: `?.` and stuff like that is not a good idea in these script files since they need to run in node, but I get your point.

I didn't know you could index on an empty array in JS, that's tricky but neat!

File: src/Middleware/Spa/SpaProxy/src/SpaHostingStartup.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+
+[assembly: HostingStartup(typeof(Microsoft.AspNetCore.SpaProxy.SpaHostingStartup))]
```
Comment (davidfowl) [inline src/Middleware/Spa/SpaProxy/src/SpaHostingStartup.cs:9]: Why?

File: src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                    {
+                        throw;
+                    }
+                }
+
+                _disposedValue = true;
+            }
+        }
+
+        ~SpaProxyLaunchManager()
```
Comment (davidfowl) [inline src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:198]: Why is there a finalizer?

Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs:198]: To ensure cleanup of the process in as many situations as possible. This is a development time only package

File: src/Middleware/Spa/SpaProxy/src/SpaHostingStartup.cs (hunk 13 above)
Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaHostingStartup.cs:9]: This is a development time only package that we wire-up in launchSettings.json and that takes care of launching the angular/react proxy server when you do `dotnet run` it's not used in the published app.

Comment (javiercn) [reply src/Middleware/Spa/SpaProxy/src/SpaHostingStartup.cs:9]: You can find all the details [here](https://gist.github.com/javiercn/b28cb0ea0912de8313e0fd0cff9021d9#file-spec-md)
